clients/src/main/java/org/apache/kafka/common/security/kerberos/Login.java

```diff
@@ -0,0 +1,381 @@
+                                            + " " + kinitArgs + "'" + "; exception was:" + e + ". Exiting refresh thread.", e);
+                                    return;
+                                }
+                            }
+                        }
+                    }
+                    try {
+                        int retry = 1;
+                        while (retry >= 0) {
+                            try {
+                                reLogin();
+                                break;
+                            } catch (LoginException le) {
+                                if (retry > 0) {
+                                    --retry;
+                                    // sleep for 10 seconds.
+                                    try {
+                                        Thread.sleep(10 * 1000);
+                                    } catch (InterruptedException e) {
+                                        log.error("Interrupted during login retry after LoginException:", le);
+                                        throw le;
+                                    }
+                                } else {
+                                    log.error("Could not refresh TGT for principal: " + principal + ".", le);
+                                }
+                            }
+                        }
+                    } catch (LoginException le) {
+                        log.error("Failed to refresh TGT: refresh thread exiting now.", le);
+                        return;
+                    }
+                }
+            }
+        }, false);
+    }
+
+    public void startThreadIfNeeded() {
+        // thread object 't' will be null if a refresh thread is not needed.
+        if (t != null) {
+            t.start();
+        }
+    }
+
+    public void shutdown() {
+        if ((t != null) && (t.isAlive())) {
+            t.interrupt();
+            try {
+                t.join();
+            } catch (InterruptedException e) {
+                log.warn("error while waiting for Login thread to shutdown: " + e);
+            }
+        }
+    }
+
+    public Subject subject() {
+        return subject;
+    }
+
+    public String loginContextName() {
+        return loginContextName;
+    }
+
+    private synchronized LoginContext login(final String loginContextName) throws LoginException {
+        if (System.getProperty(JaasUtils.JAVA_LOGIN_CONFIG_PARAM) == null) {
+            throw new IllegalArgumentException("You must pass " + JaasUtils.JAVA_LOGIN_CONFIG_PARAM + " in secure mode.");
+        }
+
+        File configFile = new File(System.getProperty(JaasUtils.JAVA_LOGIN_CONFIG_PARAM));
+        try {
+            Configuration loginConf = Configuration.getInstance("JavaLoginConfig", new URIParameter(configFile.toURI()));
+            Configuration.setConfiguration(loginConf);
```

Comment:
Is there a reason why this isn't simply using `Configuration.getConfiguration()` to get the default configuration since it is using the standard Java property to get the Jaas config file anyway? I think `JavaLoginConfig` is provided by the Sun provider, dont think it is available with all vendors.